File: java-bigdoors/src/main/java/com/bigdoors/domain/PanelRotation.java
package com.bigdoors.domain;

import com.bigdoors.util.Constants;

/**
 * Computes rotation state values for panel blocks based on door configuration.
 * Pure domain class — no Minecraft imports allowed.
 */
public final class PanelRotation {

    private PanelRotation() {}

    /**
     * Returns the panel_rotation value (0-7) for a closed panel.
     *
     * @param doorSide direction from hinge to panel (e.g. "east", "up")
     * @param facing   assembly facing direction
     * @param geoClass geometry class (0=full block, or GEOMETRY_CLASS_FENCE/BARS/SLAB/PANE)
     */
    public static int closedRotation(String doorSide, String facing, int geoClass) {
        if ("up".equals(doorSide) || "down".equals(doorSide)) {
            boolean isExotic = geoClass == Constants.GEOMETRY_CLASS_FENCE
                    || geoClass == Constants.GEOMETRY_CLASS_SLAB
                    || geoClass == Constants.GEOMETRY_CLASS_BARS
                    || geoClass == Constants.GEOMETRY_CLASS_PANE;
            if (isExotic) {
                if ("north".equals(facing) || "south".equals(facing)) return 2;
                return 1;
            }
            if ("north".equals(facing) || "south".equals(facing)) return 5;
            return 4;
        }
        return switch (doorSide) {
            case "east" -> 2;
            case "south" -> 1;
            case "west" -> 0;
            default -> 3; // north
        };
    }

    /**
     * Returns the panel_rotation value (0-7) for an open panel.
     *
     * @param mode      "horizontal" or "vertical"
     * @param doorSide  direction from hinge to panel
     * @param facing    assembly facing direction
     * @param direction "cw" or "ccw"
     * @param geoClass  geometry class
     */
    public static int openRotation(String mode, String doorSide, String facing, String direction, int geoClass) {
        if ("vertical".equals(mode)) {
            boolean isExotic = geoClass == Constants.GEOMETRY_CLASS_FENCE
                    || geoClass == Constants.GEOMETRY_CLASS_SLAB
                    || geoClass == Constants.GEOMETRY_CLASS_BARS
                    || geoClass == Constants.GEOMETRY_CLASS_PANE;
            if (isExotic) {
                if ("north".equals(facing) || "south".equals(facing)) return 6;
                return 5;
            }
            if ("north".equals(facing) || "south".equals(facing)) return 1;
            return 0;
        }
        int closed = closedRotation(doorSide, facing, geoClass);
        return "cw".equals(direction) ? (closed + 3) % 4 : (closed + 1) % 4;
    }
}

File: bigdoors_bp/scripts/domain/PanelRotation.js
import {
  GEOMETRY_CLASS_FENCE,
  GEOMETRY_CLASS_SLAB,
  GEOMETRY_CLASS_BARS,
  GEOMETRY_CLASS_PANE,
} from "../util/Constants.js";

export function closedRotation(doorSide, facing, geoClass) {
  if (doorSide === "up" || doorSide === "down") {
    if (geoClass === GEOMETRY_CLASS_FENCE || geoClass === GEOMETRY_CLASS_SLAB || geoClass === GEOMETRY_CLASS_BARS || geoClass === GEOMETRY_CLASS_PANE) {
      if (facing === "north" || facing === "south") return 2;
      return 1;
    }
    if (facing === "north" || facing === "south") return 5;
    return 4;
  }
  if (doorSide === "east") return 2;
  if (doorSide === "south") return 1;
  if (doorSide === "west") return 0;
  return 3;
}

export function openRotation(mode, doorSide, facing, direction, geoClass) {
  if (mode === "vertical") {
    if (geoClass === GEOMETRY_CLASS_FENCE || geoClass === GEOMETRY_CLASS_SLAB || geoClass === GEOMETRY_CLASS_BARS || geoClass === GEOMETRY_CLASS_PANE) {
      if (facing === "north" || facing === "south") return 6;
      return 5;
    }
    if (facing === "north" || facing === "south") return 1;
    return 0;
  }
  const closed = closedRotation(doorSide, facing, geoClass);
  return direction === "cw" ? (closed + 3) % 4 : (closed + 1) % 4;
}
